File: hmis-engine-api/src/main/java/com/otapp/hmis/engine/masterdata/pharmacy/application/PharmacyService.java
package com.otapp.hmis.engine.masterdata.pharmacy.application;

import com.otapp.hmis.engine.common.api.PageResponse;
import com.otapp.hmis.engine.common.error.ConflictException;
import com.otapp.hmis.engine.common.error.NotFoundException;
import com.otapp.hmis.engine.masterdata.pharmacy.application.PharmacyDtos.CreatePharmacyRequest;
import com.otapp.hmis.engine.masterdata.pharmacy.application.PharmacyDtos.PharmacyDto;
import com.otapp.hmis.engine.masterdata.pharmacy.application.PharmacyDtos.UpdatePharmacyRequest;
import com.otapp.hmis.engine.masterdata.pharmacy.domain.Pharmacy;
import com.otapp.hmis.engine.masterdata.pharmacy.domain.PharmacyRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class PharmacyService {

    private final PharmacyRepository pharmacyRepository;

    @Transactional
    public PharmacyDto create(CreatePharmacyRequest request) {
        String code = request.code().trim().toUpperCase();
        if (pharmacyRepository.existsByCode(code)) {
            throw new ConflictException("Pharmacy code already exists: " + code);
        }
        Pharmacy p = new Pharmacy(code, request.name().trim(), request.location(), request.description());
        pharmacyRepository.save(p);
        return toDto(p);
    }

    @Transactional
    public PharmacyDto update(String uid, UpdatePharmacyRequest request) {
        Pharmacy p = loadOrThrow(uid);
        p.setName(request.name().trim());
        p.setLocation(request.location());
        p.setDescription(request.description());
        return toDto(p);
    }

    @Transactional
    public PharmacyDto setActive(String uid, boolean active) {
        Pharmacy p = loadOrThrow(uid);
        if (active) p.activate(); else p.deactivate();
        return toDto(p);
    }

    @Transactional
    public void delete(String uid) {
        pharmacyRepository.delete(loadOrThrow(uid));
    }

    @Transactional(readOnly = true)
    public PharmacyDto findByUid(String uid) {
        return toDto(loadOrThrow(uid));
    }

    @Transactional(readOnly = true)
    public PageResponse<PharmacyDto> search(String query, Boolean active, Pageable pageable) {
        return PageResponse.from(
                pharmacyRepository.search(query == null ? null : query.trim(), active, pageable).map(PharmacyService::toDto));
    }

    private Pharmacy loadOrThrow(String uid) {
        return pharmacyRepository.findByUid(uid).orElseThrow(() -> new NotFoundException("Pharmacy not found: " + uid));
    }

    private static PharmacyDto toDto(Pharmacy p) {
        return new PharmacyDto(p.getUid(), p.getCode(), p.getName(), p.getLocation(),
                p.getDescription(), p.isActive(), p.getCreatedAt(), p.getUpdatedAt());
    }
}

File: hmis-engine-web/src/app/features/masterdata/pharmacies/pharmacy.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';

import { environment } from '../../../../environments/environment';
import { PageResponse } from '../../../core/http/page.types';
import { CreatePharmacyRequest, Pharmacy, PharmacySearchParams, UpdatePharmacyRequest } from './pharmacy.types';

@Injectable({ providedIn: 'root' })
export class PharmacyService {
  private readonly http = inject(HttpClient);
  private readonly base = `${environment.apiUrl}/masterdata/pharmacies`;

  search(params: PharmacySearchParams = {}): Observable<PageResponse<Pharmacy>> {
    let p = new HttpParams();
    if (params.query) p = p.set('query', params.query);
    if (params.active !== undefined) p = p.set('active', String(params.active));
    if (params.page !== undefined) p = p.set('page', String(params.page));
    if (params.size !== undefined) p = p.set('size', String(params.size));
    if (params.sort) p = p.set('sort', params.sort);
    return this.http.get<PageResponse<Pharmacy>>(this.base, { params: p });
  }

  findByUid(uid: string): Observable<Pharmacy> { return this.http.get<Pharmacy>(`${this.base}/${uid}`); }
  create(req: CreatePharmacyRequest): Observable<Pharmacy> { return this.http.post<Pharmacy>(this.base, req); }
  update(uid: string, req: UpdatePharmacyRequest): Observable<Pharmacy> { return this.http.put<Pharmacy>(`${this.base}/${uid}`, req); }
  setActive(uid: string, active: boolean): Observable<Pharmacy> { return this.http.put<Pharmacy>(`${this.base}/${uid}/active`, { active }); }
  delete(uid: string): Observable<void> { return this.http.delete<void>(`${this.base}/${uid}`); }
}
